colossalai/inference/quant/smoothquant/models/smoothquant_layer.py

```diff
@@ -0,0 +1,185 @@
+# Code modified from smoothquant: https://github.com/mit-han-lab/smoothquant
+
+from typing import Optional, Tuple
+
+import torch
+from torch import nn
+from torch_int.nn.bmm import BMM_S8T_S8N_F32T, BMM_S8T_S8N_S8T
+from torch_int.nn.linear import W8A8B8O8Linear, W8A8BFP32OFP32Linear
+from transformers.models.llama.modeling_llama import LlamaAttention
+
+from colossalai.kernel.triton import int8_rotary_embedding_fwd
+
+
+class LLamaSmoothquantAttention(nn.Module):
+    def __init__(
+        self,
+        hidden_size: int,
+        num_heads: int,
+    ):
+        super().__init__()
+        self.hidden_size = hidden_size
+        self.num_heads = num_heads
+        self.head_dim = hidden_size // num_heads
+
+        if (self.head_dim * num_heads) != self.hidden_size:
+            raise ValueError(
+                f"hidden_size must be divisible by num_heads (got `hidden_size`: {self.hidden_size}"
+                f" and `num_heads`: {num_heads})."
+            )
+
+        self.attention_weight_scale = 1.0
+
+        self.qk_bmm = BMM_S8T_S8N_F32T(1.0)
+        self.pv_bmm = BMM_S8T_S8N_S8T(1.0)
+
+        self.k_proj = W8A8B8O8Linear(hidden_size, hidden_size)
+        self.v_proj = W8A8B8O8Linear(hidden_size, hidden_size)
+        self.q_proj = W8A8B8O8Linear(hidden_size, hidden_size)
+        self.out_proj = W8A8BFP32OFP32Linear(hidden_size, hidden_size)
+
+        self.q_output_scale = torch.tensor([1.0])
+        self.k_output_scale = torch.tensor([1.0])
+        self.rotary_output_scale = torch.tensor([1.0])
+
+    def pack(
```

Comment:
Is the function required in future ? (seems not to be used in this PR) What 's the purpose of the function

Reply:
It used to get int8 weight and other values from original models.